File: 240. Search a 2D Matrix II/240. Search a 2D Matrix II.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int right = matrix[0].size()-1;
        for(const auto& row: matrix){
            if(row[0] > target)
                return false;
            int l = 0, r = right, mid;
            while(l <= r){
                mid = l+(r-l)/2;
                if(row[mid] == target)
                    return true;
                if(row[mid] > target)
                    r = mid-1;
                else
                    l = mid+1;
            }
            right = max({l,r,mid})-1;
        }
        return false;
    }
};

File: 240. Search a 2D Matrix II/240. Search a 2D Matrix II.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int m = matrix.length, n = matrix[0].length, r = 0, c = n-1;
        while(r < m && c >= 0){
            if(matrix[r][c] < target)   ++r;
            else if(matrix[r][c] > target)  --c;
            else    return true;
        }
        return false;
    }
}